File: backend/app/hms/pms/schemas/reports.py
from datetime import date
from enum import Enum

from pydantic import BaseModel, Field

from app.hms.schemas import HotelReportDailyRead, HotelReportSummaryRead


class PmsReportType(str, Enum):
    COCKPITLISTE = "cockpitliste"
    HOUSEKEEPINGLISTE = "housekeepingliste"
    HAUS_STATUS = "haus_status"
    FB_VERPFLEGUNGSBERICHT = "fb_verpflegungsbericht"
    KASSENBUCH = "kassenbuch"
    ANZAHLUNGSLISTE = "anzahlungsliste"
    EINNAHMEBERICHT = "einnahmebericht"
    FINANZKONTEN_UEBERSICHT = "finanzkonten_uebersicht"
    OFFENE_SALDEN = "offene_salden"
    RECHNUNGSBERICHT = "rechnungsbericht"
    WARENGRUPPENJOURNAL = "warengruppenjournal"
    BELEGUNGSUEBERSICHT = "belegungsuebersicht"
    TAGESZAHLEN = "tageszahlen"
    BUCHUNGSQUELLENBERICHT = "buchungsquellenbericht"
    KENNZAHLENBERICHT = "kennzahlenbericht"
    CITY_TAX_BERICHT = "city_tax_bericht"
    GOBD_EXPORT = "gobd_export"
    MELDESCHEIN_DOWNLOAD = "meldeschein_download"
    FREMDENVERKEHRSSTATISTIK_XML = "fremdenverkehrsstatistik_xml"


class PmsReportDownloadQuery(BaseModel):
    type: PmsReportType
    start: date | None = None
    end: date | None = None
    property_id: int | None = Field(default=None, gt=0)


class PmsReportSummaryRead(HotelReportSummaryRead):
    pass


class PmsReportDailyRead(HotelReportDailyRead):
    pass

File: frontend/src/features/hms/pms/api/reports.ts
"use client";

import api, { getJson } from "@/lib/api";
import { defaultHotelPropertyId } from "@/lib/hotel-room-types";
import type { HotelReportDaily, HotelReportSummary } from "@/lib/hms";

export type PmsReportType =
  | "cockpitliste"
  | "housekeepingliste"
  | "haus_status"
  | "fb_verpflegungsbericht"
  | "kassenbuch"
  | "anzahlungsliste"
  | "einnahmebericht"
  | "finanzkonten_uebersicht"
  | "offene_salden"
  | "rechnungsbericht"
  | "warengruppenjournal"
  | "belegungsuebersicht"
  | "tageszahlen"
  | "buchungsquellenbericht"
  | "kennzahlenbericht"
  | "city_tax_bericht"
  | "gobd_export"
  | "meldeschein_download"
  | "fremdenverkehrsstatistik_xml";

function filenameFromDisposition(headerValue: string | undefined, fallback: string) {
  if (!headerValue) return fallback;
  const match = /filename="?([^"]+)"?/i.exec(headerValue);
  return match?.[1] || fallback;
}

export async function fetchPmsReportSummary(propertyId: number = defaultHotelPropertyId, days = 30) {
  return getJson<HotelReportSummary>("/hms/pms/reports/summary", {
    params: { property_id: propertyId, days },
  });
}

export async function fetchPmsReportDaily(propertyId: number = defaultHotelPropertyId, days = 30) {
  return getJson<HotelReportDaily>("/hms/pms/reports/daily", {
    params: { property_id: propertyId, days },
  });
}

export async function downloadPmsReport(
  reportType: PmsReportType,
  {
    propertyId = defaultHotelPropertyId,
    start,
    end,
  }: {
    propertyId?: number;
    start?: string;
    end?: string;
  } = {},
) {
  const response = await api.get<Blob>("/hms/pms/reports/download", {
    params: {
      type: reportType,
      property_id: propertyId,
      start,
      end,
    },
    responseType: "blob",
  });

  const contentType = response.headers["content-type"] || "application/octet-stream";
  const blob = new Blob([response.data], { type: contentType });
  const fallbackExtension = reportType === "fremdenverkehrsstatistik_xml" ? "xml" : "csv";
  const fallbackName = `${reportType}.${fallbackExtension}`;
  const filename = filenameFromDisposition(response.headers["content-disposition"], fallbackName);
  const href = URL.createObjectURL(blob);
  const link = document.createElement("a");
  link.href = href;
  link.download = filename;
  document.body.appendChild(link);
  link.click();
  link.remove();
  window.setTimeout(() => URL.revokeObjectURL(href), 1000);
}
